src/Http/Http.Extensions/gen/RequestDelegateGenerator.cs

```diff
@@ -79,6 +79,7 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
             codeWriter.StartBlock();
             codeWriter.WriteLine(@"Debug.Assert(options?.EndpointBuilder != null, ""EndpointBuilder not found."");");
             codeWriter.WriteLine($"options.EndpointBuilder.Metadata.Add(new SourceKey{endpoint!.EmitSourceKey()});");
+            endpoint!.EmitEndpointMetadataPopulation(codeWriter);
```

Comment:
nit: The bang on `endpoint!` is repeated _everywhere_. I was going to suggest adding `Debug.Assert`, but it looks like you could replace the where on line 56 with an OfType to stop it being nullable everywhere.

For example:

```diff
- .Where(static endpoint => endpoint != null)
+ .OfType<Endpoint>() 
```

Then remove the bangs everywhere.

Reply:
That `Where(...)` is not operating on a enumerable/collection (confusingly). Same for `Select(...)`. Talk about choosing some surprising method names and return types ;)

Anyway - I think Debug.Assert is a fine suggestion ;)

Reply:
Thinking about this some more, I am going to do a follow up PR for this. There are a few places where we do things like spin over the entire collection of endpoints multiple times in a single method. I want to optimize this and as I do it address a of the maybe nulls.

Reply:
For tracking: https://github.com/dotnet/aspnetcore/issues/47771

I've put it in preview 5, but hopefully we can get some tweaks in before we snap preview 4.